ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -0,0 +1,187 @@
+`shp-error-reason` have a recognizable name that is ideally prefixed
+with the problematic technology, for example, `git` in case of missing
+a secret. The result `shp-error-message` should be descriptive text to
+give the user enough information to troubleshoot the BuildRun.  A task
+run emits the new results as Tekton results.  The BuildRun CRD is
+assumed to have the `status.results` according to `0023` (alternatives
+are discussed further below). 
+Finally, we must extend the BuildRun controller
+logic to propagate the new results into the BuildRun CRD's
+`status.results` property.
+
+BuildRun with new error results: 
+
+```yaml
+status:
+   results:
+       - name: shp-error-reason
+         value: git-ssh-auth
+       - name: shp-error-message
+         value: Authentication failed for ...
+```
+
+We do not suggest to mutate the `status.conditions.message` field
+since it contains valuable hints to troubleshoot on a lower level if
+the propagated error message and reason is not sufficient.
+
+### User Stories
+
+#### Story 1 - Developer and direct user of shipwright perspective
+As a user of Shipwright and developer, I want to identify why my
+BuildRun is in a failed state without going through the lengthy
+process of identifying the correct container and manually processing
+its logs.
+#### Story 2 - Github Actions developer (third party application developer)
+As a third-party application developer that uses shipwright under the hood, I want to supply my users with meaningful hints to fix their builds, thus requiring a straightforward API to consume error details.
+
+### Implementation Notes
+
+`tbd in the next phase of the proposal`
+
+### Test Plan
+
+`tbd in the next phase of the proposal`
+
+### Release Criteria
+
+`tbd in the next phase of the proposal`
+
+### Risks and Mitigations
+Tekton results are limited in size (4096 bytes) which applies to the
+proposed error results.
+
+There is no strict typing that would prevent strategy authors from
+using proper error reasons. So two authors can use different names for
+the same error reason or use the same name for different errors. To
+prevent these issues, we need clear documentation with well-known
+error reasons.
+
+Ideally, this ship requires that TaskRuns emit results in error
+cases. However, Tekton currently does not emit any results for failed
+TaskRuns. Therefore, I have reached out to the Tekton community to
+include that functionality
+(https://github.com/tektoncd/pipeline/issues/3749).  For the case that
+the request is denied, for example, because of Tekton's API design
+philosophy, I have researched a workaround. In every TaskRun step's
+status, there is a message property (JSON string), including the step
+results. It requires iterating all steps and processing the message
+field.
+
+# Drawbacks
+Since we only expose a field for an error's message and reason, we do
+not allow strategy authors to introduce additional custom fields for
+clarifying the build problem.
```

Comment:
Results would still not be emitted in case of errors in the TaskRun. We would have to recognise the strategy author's additional field to explicitly lift them up into our buildrun.

So the feature by itself is not enough to mitigate the drawback.

Reply:
True.